homeassistant/components/mealie/todo.py

```diff
@@ -0,0 +1,260 @@
+
+from homeassistant.components.todo import (
+    TodoItem,
+    TodoItemStatus,
+    TodoListEntity,
+    TodoListEntityFeature,
+)
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .coordinator import MealieConfigEntry, MealieShoppingListCoordinator
+from .entity import MealieEntity
+
+TODO_STATUS_MAP = {
+    False: TodoItemStatus.NEEDS_ACTION,
+    True: TodoItemStatus.COMPLETED,
+}
+TODO_STATUS_MAP_INV = {v: k for k, v in TODO_STATUS_MAP.items()}
+
+
+def _convert_api_item(item: ShoppingItem) -> TodoItem:
+    """Convert Mealie shopping list items into a TodoItem."""
+
+    return TodoItem(
+        summary=item.display,
+        uid=item.item_id,
+        status=TODO_STATUS_MAP.get(
+            item.checked,
+            TodoItemStatus.NEEDS_ACTION,
+        ),
+        due=None,
+        description=None,
+    )
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: MealieConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the todo platform for entity."""
+    coordinator = entry.runtime_data.shoppinglist_coordinator
+
+    async_add_entities(
+        MealieShoppingListTodoListEntity(coordinator, shopping_list)
+        for shopping_list in coordinator.shopping_lists
+    )
+
+
+class MealieShoppingListTodoListEntity(MealieEntity, TodoListEntity):
+    """A todo list entity."""
+
+    _attr_supported_features = (
+        TodoListEntityFeature.CREATE_TODO_ITEM
+        | TodoListEntityFeature.UPDATE_TODO_ITEM
+        | TodoListEntityFeature.DELETE_TODO_ITEM
+        | TodoListEntityFeature.MOVE_TODO_ITEM
+    )
+
+    coordinator: MealieShoppingListCoordinator
+
+    def __init__(
+        self, coordinator: MealieShoppingListCoordinator, shopping_list: ShoppingList
+    ) -> None:
+        """Create the todo entity."""
+        super().__init__(coordinator, shopping_list.list_id)
+        self._shopping_list = shopping_list
+        self._attr_name = shopping_list.name
+        self._attr_unique_id = (
+            f"{self.coordinator.config_entry.entry_id}_{shopping_list.list_id}"
+        )
+        self.translation_key = "shopping_list"
+
+    @property
+    def todo_items(self) -> list[TodoItem] | None:
+        """Get the current set of To-do items."""
+        if self._shopping_list.list_id in self.coordinator.shopping_list_items:
+            return [
+                _convert_api_item(item)
+                for item in self.coordinator.shopping_list_items[
+                    self._shopping_list.list_id
+                ]
+            ]
+
+        return []
+
+    async def async_create_todo_item(self, item: TodoItem) -> None:
+        """Add an item to the list."""
+        position = 0
+        if len(self.coordinator.shopping_list_items[self._shopping_list.list_id]) > 0:
```

Comment:
Instead of referencing this every time, maybe it makes sense to have a
```py
@property
def shopping_items(self) -> list[..]:
    return self.coordinator.shopping_list_items[self._shopping_list.list_id]
```
So you can easily access it via `self.shopping_items`